hydroflow/examples/kvs/kvs_compiled.rs

```diff
@@ -92,50 +93,162 @@ impl Iterator for PerQuantumPulser {
+        + Clone,
+    K: 'static + Send + Clone,
+    V: 'static + Ord + Send + Clone,
+{
+    let mut core_ids = core_affinity::get_core_ids().unwrap();
+
+    let (monitor_send, monitor_recv) = channel(1024);
+
+    let (matrix, senders) = make_communication_matrix(n);
+    for (i, (receiver, senders)) in matrix.into_iter().enumerate() {
+        let f = f.clone();
+        let id = core_ids.pop().unwrap();
+        let monitor = monitor_send.clone();
+        let config = config.clone();
+        std::thread::spawn(move || {
+            core_affinity::set_for_current(id);
+            f(i, receiver, senders, monitor, config);
+        });
+    }
+
+    (senders, monitor_recv)
+}
+
+#[derive(Clone)]
+pub struct WorkerConfig {
+    pub epoch: Duration,
+    pub batch_size: usize,
+    pub dist: Dist,
+}
+
+fn spawn_threads<K, V>(
+    workers: u64,
+    config: WorkerConfig,
+) -> (Vec<Sender<Message<K, V>>>, Receiver<MonitorInfo>)
+where
+    K: 'static + Clone + Eq + std::hash::Hash + Send + std::fmt::Debug + Default,
     V: 'static + Clone + Send + std::fmt::Debug + Ord + Default,
+    Dist: Sample<K>,
 {
     spawn(
         workers,
-        move |id, receiver: Receiver<Message<K, V>>, mut senders: Vec<Sender<Message<K, V>>>| {
+        config,
+        move |id,
+              _receiver: Receiver<Message<K, V>>,
+              mut senders: Vec<Sender<Message<K, V>>>,
+              mut monitor: Sender<MonitorInfo>,
+              config: WorkerConfig| {
             let rt = tokio::runtime::Runtime::new().unwrap();
+            let mut dist = config.dist.clone();
             rt.block_on(async move {
                 let mut hf = HydroflowBuilder::default();
 
                 let (z_send, z_recv) = hf.add_channel_input::<_, Option<u64>, VecHandoff<u64>>("ticks");
 
                 // Construct the ticker.
-                let epoch_duration = Duration::from_millis(100);
                 tokio::spawn(async move {
                     let mut tick = 0;
                     loop {
-                        tokio::time::sleep(epoch_duration).await;
+                        tokio::time::sleep(config.epoch).await;
                         z_send.give(Some(tick));
                         z_send.flush();
                         tick += 1;
                     }
                 });
 
-                let q_recv = hf.add_input_from_stream::<_, Option<_>, VecHandoff<_>, _>("writes", receiver.map(Some));
+                let (system_send, system_recv) = hf.make_edge::<_, VecHandoff<Message<K, V>>, VecDeque<Message<K, V>>>("writes");
+
+                let mut rng = rand::thread_rng();
+
+                // Once per epoch, generate a batch of data.
+                hf.add_subgraph(
+                    "workload_generator",
+                    IterPullSurface::new(PerQuantumPulser::new()).map_with_context(|ctx, _| {
+                        ctx.waker().wake();
+                        ctx.current_epoch()
+                    }).map(move |_|{
```

Comment:
Probably put a TODO here for me to think about how we might make this and other input generators cleaner